<a href="https://colab.research.google.com/github/JuanManuelLopezP/Anal-tica/blob/main/RETO_2_Juan_Manuel_L%C3%B3pez_Pe%C3%B1a.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier


from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
from sre_constants import ANY_ALL
nxl='/content/drive/MyDrive/Analítica/RETO 2.xlsx'
XDB=pd.read_excel(nxl, sheet_name=1)
XDB=XDB.dropna()
XDB.head(100)


XD=np.array(XDB[['College_','FoodSpend','Income','TravelSpend']])
print(XD)
yd=np.array(XDB['Credit_Card'])
print(yd)

[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[1 1 1 1 1 0 0 1 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1 0 1 0 1 1 0 1 0 1 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 1
 0 0 1 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 1 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 0 0 0
 1 0 1 0 0 1 1 1 0 0 0 1 1 0 0 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 0 0 0 1 0 0 0
 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 1 1 0 0 0 1 0 0 1 0 0 0 0 0
 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 0 1 0 0 1 0 1 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0
 0 0 0 0 0 0 1 1 1 1 0 1 1 1 0 1 1 1 1 1 0 0 0 1 0 0 0 0 1 1 1 0 0 1 0 1 0
 0 0 0 0 0 1 

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar=mar.fit(XD,yd)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

#Evaluamos modelo
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

Ex=(VN+VP)/(VN+VP+FN+FP)   #Que porcentage de los datos pronosticó correctamente
print("La exactitud alcanzada por el modelo es:",Ex)
se=VP/(FN+VP) #Como le fue con los positivos
print("La sensibildiad alcanzada por el modelo fue:",se)
esp=VN/(VN+FP) #Como le fue con los negativos
print("La especificidad alcanzada por el modelo fue:",esp)

[[334   9]
 [ 42 115]]
La exactitud alcanzada por el modelo es: 0.898
La sensibildiad alcanzada por el modelo fue: 0.732484076433121
La especificidad alcanzada por el modelo fue: 0.9737609329446064


In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College_','FoodSpend','Income','TravelSpend'] #Son titulos del arbol
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol2.png')

True

In [ ]:
XDB2=pd.read_excel(nxl, sheet_name=1)
XDB2=XDB2.dropna()
XDB2.head(100)

XDB2=np.array(XDB2[['College_','FoodSpend','Income','TravelSpend']]) #Datos entrada
print(XDB2)

#Pronosticamos la Preaprobación
ydp2=mar.predict(XDB2)
print(ydp2)
print("El número de personas que usaran la tarjeta es de :",np.sum(ydp2))
print("El número total de datos es:",len(ydp2))
print("El número de personas que no usaran la tarjeta es de:",len(ydp2)-np.sum(ydp2))

[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[1 1 1 1 1 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 1 0 1 0 0 1 0 1 0 0 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 1 0 0 0 1 0 0 1 0 1 0 0 0 0 0 0 1
 0 0 1 0 0 0 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0
 0 1 1 0 0 1 1 1 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0
 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 1 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0
 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0
 0 0 0 0 0 0 0 0 1 1 0 1 1 1 0 0 1 1 0 1 0 0 0 1 0 0 0 0 1 1 0 0 0 1 0 1 0
 0 0 0 0 0 0 

**Analisis de resultados**

Con el modelo anteriormente planteado, podemos observar que este mismo presenta una buena clasificación para determinar si si una persona usará su tarjeta de crédito o no a la hora de ir de viaje, esto gracias a que el modelo cuenta con una exactitud de 89% del modelo en general, para la sensibilidad podemos determinar que el modelo alcanza un 73.2%, la cual permite categorizar correctamente aquellas personas que si usarán su tarjeta de crédito, y la especifidad alcanzó un 97.37% para clasificar correctamente a las personas que no usaran su tarjeta de crédito para pagar su viaje.

Con los nodos puros, podemos encontrar 3 de estos los cuales son:

  [1,0] (SI TravelSpend<=1188 AND FoodSpend>4972.55)

  [0,47] (SI TravelSpend<=1188 AND FoodSpend>4990.73)

  [108,0] (SI TravelSpend>1567.875 AND Income>59098)

Para el caso de los nuevos clientes podemos ver que estos presentan las siguientes categorías

**Primer cliente**: Para este, se presenta la siguiente regla: SI TravelSpend>1567.875 AND Income 59098, la cual nos deja con un nodo puro de [108,0] lo cual significa que este cliente tiene una posibilidad del 100% de usar su tarjeta de crédito para pagar su viaje.

**Segundo Cliente**: Este presenta la siguiente regla: SI Income < 59098 AND FoodSpend > 2711.895 AND TravelSpend < 3008.25, esta persona tiene una posibilidad del 78.67% de usar su tarjeta de crédito para pagar su viaje